### PR TITLE
Respond correctly to unknown methods on known endpoints

### LINKED ISSUE #14599: Synapse returns improper error codes for unknown endpoints
As per [MSC3743](https://github.com/matrix-org/matrix-spec-proposals/pull/3743). We also don't properly handle homeservers which *do* return the proper error codes when failing over.

### PR DESCRIPTION
**Pull request by clokep**
Respond with a `405` error if a request is received on a known endpoint, but to an unknown method. See MSC3743.

Should be reviewable commit-by-commit.

Fixes #14599 

Complement tests at matrix-org/complement#565

### CODE REVIEW

**Comment by clokep** [comment]
Arg this is failing because we now have conflicting paths:

* `^/_synapse/admin/v1/media/(?P<server_name>[^/]*)/(?P<media_id>[^/]*)$`
* `^/_synapse/admin/v1/media/(?P<server_name>[^/]*)/delete$`

**Comment by clokep** [comment]
I split out everything except the routing changes to separate PRs (#14620, #14621). Those should be straightforward and non-controversial.

**Comment by clokep** [comment]
I'm throwing this on the queue. There's some CI failures, but I don't think they're related.

What do folks think of this change? Is it acceptable? It is a real shame that our routing logic is such a mess. 😢

**Review by squahtx — APPROVED** [review]
Looks very reasonable!

File: docs/admin_api/media_admin_api.md [hunk 1]
```diff
@@ -235,6 +235,14 @@ The following fields are returned in the JSON response body:
 
 Request:
 
+```
+POST /_synapse/admin/v1/media/delete?before_ts=<before_ts>
+
+{}
+```
+
+*Deprecated in Synapse v1.77.0:* This API is available at the deprecated endpoint:
```
Comment (clokep) [inline docs/admin_api/media_admin_api.md:244]:
```suggestion
*Deprecated in Synapse v1.78.0:* This API is available at the deprecated endpoint:
```

File: docs/upgrade.md [hunk 2]
```diff
@@ -88,6 +88,15 @@ process, for example:
     dpkg -i matrix-synapse-py3_1.3.0+stretch1_amd64.deb
     ```
 
+# Upgrading to v1.77.0
```
Comment (clokep) [inline docs/upgrade.md:91]:
```suggestion
# Upgrading to v1.78.0
```

File: docs/upgrade.md [hunk 3]
```diff
@@ -88,6 +88,15 @@ process, for example:
...
+
+## Deprecate the `/_synapse/admin/v1/media/<server_name>/delete` admin API
+
+Synapse 1.77.0 replaces the `/_synapse/admin/v1/media/<server_name>/delete`
```
Comment (clokep) [inline docs/upgrade.md:95]:
```suggestion
Synapse 1.78.0 replaces the `/_synapse/admin/v1/media/<server_name>/delete`
```

**Comment by dklimpel** [comment]
Fixes parts of #9284

**Pull request merged**